Matrix-Free Linear Operator

e-moral-sanchez opened the pull request:
This PR provides a `MatrixFreeLinearOperator` class which is a subclass of the abstract class `LinearOperator`.

The new class allows to create a general matrix-free linear operator. The constructor only requires the domain, codomain and a callable `dot` method.

Notes:
- the provided `dot` method may or may not take an 'out' argument
- a `transpose_dot` method may also be provided (mandatory to instantiate the transpose() linear operator)
- unit tests have been added

Additional change: 
- in this PR we stop passing the deprecated `tol` argument in calls to scipy's minres. This forces us to use scipy >= 1.14 and hence python >= 3.9
- we also added a convergence verification in minres linear solver to avoid iterating when the initial solution is good enough (which may cause a division-by-zero error)

e-moral-sanchez committed:
simple implementation GeneralLinearOperator

e-moral-sanchez committed:
Merge branch 'devel' into general_linear_operator

e-moral-sanchez committed:
general linear operator needs toarray, tosparse,transpose method

e-moral-sanchez committed:
improve docstrings

e-moral-sanchez committed:
Merge branch 'devel' into general_linear_operator

e-moral-sanchez committed:
fixed message notimplementederror

e-moral-sanchez committed:
put import outside class

campospinto commented:
This is useful I think, thanks! 

As a test, you can check the behaviour of a "fake" matrix-free operator: an operator created with a dot product involving a Stencil Matrix should act exactly like the matrix itself.

You can also check that an iterative solver (like cg) can run with an object of this new class.

campospinto commented:
I see that some tests are failing because of `minres()` getting an unexpected keyword argument `tol`. Is this related to this new class?

e-moral-sanchez commented:
The new version `1.14.0`  of SciPy removed the `tol` argument in `minres`: https://docs.scipy.org/doc/scipy-1.13.1/reference/generated/scipy.sparse.linalg.minres.html 

This also affects PR https://github.com/campospinto/psydac_dev/pull/6 . 

I will fix this in this PR.

e-moral-sanchez committed:
set tol to rtol in MINRES

campospinto commented on psydac/linalg/basic.py:
```diff
@@ -1111,3 +1113,52 @@ def solve(self, rhs, out=None):
+class MatrixFreeLinearOperator(LinearOperator):
+    """
+    General operator acting between two vector spaces V and W. It only requires a callable dot method.
+
```
an example of instantiation would be useful

campospinto committed:
add features and tests to MatrixFree linear ops

campospinto commented:
> This PR provides the class `MatrixFreeLinearOperator`, which is a subclass of the abstract class `LinearOperator`.
> 
> The class `MatrixFreeLinearOperator` allows to create a general matrix-free linear operator. The constructor only requires the domain, codomain and a callable `dot` method.
> 
> TODO: add simple unit tests (ideas are welcome!)

I am just adding a few unit tests

campospinto commented:
I am also adding some features to handle the case where the given dot method has not optional 'out' argument, and the possibility (not mandatory) to pass a 'transpose_dot' method

campospinto committed:
tests for matrix free linear ops

campospinto committed:
require scipy >= 1.14 in requirements.txt

Calls to scipy `minres` now use `rtol` instead of `tol` to support  version 1.14.0, but this causes previous versions of scipy to fail

campospinto committed:
discard tests with python3.8

calls to scipy's minres now only use rtol which prevents from using scipy < 1.14. In turns this prevent from using python3.8 which is close to being unsupported anyway, see https://devguide.python.org/versions/

campospinto committed:
avoid minres iteration if converged
